sw/device/tests/sim_dv/spi_tx_rx_test.c

```diff
@@ -0,0 +1,391 @@
+#include "sw/device/lib/runtime/hart.h"
+#include "sw/device/lib/testing/test_main.h"
+#include "sw/device/lib/testing/test_status.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+
+#define SPI_DEVICE_DATASET_SIZE 128
+
+static const uint16_t kFifoLen = 0x400;
+static dif_plic_t plic;
+static dif_spi_device_t spi_device;
+
+// A set of bytes to be send out by spi_device.
+static const uint8_t spi_device_tx_data[SPI_DEVICE_DATASET_SIZE] = {
+    0xe8, 0x50, 0xc6, 0xb4, 0xbe, 0x16, 0xed, 0x55, 0x16, 0x1d, 0xe6, 0x1c,
+    0xde, 0x9f, 0xfd, 0x24, 0x89, 0x81, 0x4d, 0x0d, 0x1a, 0x12, 0x4f, 0x57,
+    0xea, 0xd6, 0x6f, 0xc0, 0x7d, 0x46, 0xe7, 0x37, 0x81, 0xd3, 0x8e, 0x16,
+    0xad, 0x7b, 0xd0, 0xe2, 0x4f, 0xff, 0x39, 0xe6, 0x71, 0x3c, 0x82, 0x04,
+    0xec, 0x3a, 0x27, 0xcc, 0x3d, 0x58, 0x0e, 0x56, 0xd2, 0xd2, 0xb9, 0xa3,
+    0xb5, 0x3d, 0xc0, 0x40, 0xba, 0x90, 0x16, 0xd8, 0xe3, 0xa4, 0x22, 0x74,
+    0x80, 0xcb, 0x7b, 0xde, 0xd7, 0x3f, 0x4d, 0x93, 0x4d, 0x59, 0x79, 0x88,
+    0x24, 0xe7, 0x68, 0x8b, 0x7a, 0x78, 0xb7, 0x07, 0x09, 0x26, 0xcf, 0x6b,
+    0x52, 0xd9, 0x4c, 0xd3, 0x33, 0xdf, 0x2e, 0x0d, 0x3b, 0xab, 0x45, 0x85,
+    0xc2, 0xc2, 0x19, 0xe5, 0xc7, 0x2b, 0xb0, 0xf6, 0xcb, 0x06, 0xf6, 0xe2,
+    0xf5, 0xb1, 0xab, 0xef, 0x6f, 0xd8, 0x23, 0xfd,
+};
+
+// The set of bytes expected to be received by spi_device from spi_host.
+static const uint8_t exp_spi_device_rx_data[SPI_DEVICE_DATASET_SIZE] = {
+    0x1b, 0x95, 0xc5, 0xb5, 0x8a, 0xa4, 0xa8, 0x9f, 0x6a, 0x7d, 0x6b, 0x0c,
+    0xcd, 0xd5, 0xa6, 0x8f, 0x07, 0x3a, 0x9e, 0x82, 0xe6, 0xa2, 0x2b, 0xe0,
+    0x0c, 0x30, 0xe8, 0x5a, 0x05, 0x14, 0x79, 0x8a, 0xFf, 0x88, 0x29, 0xda,
+    0xc8, 0xdd, 0x82, 0xd5, 0x68, 0xa5, 0x9d, 0x5a, 0x48, 0x02, 0x7f, 0x24,
+    0x32, 0xaf, 0x9d, 0xca, 0xa7, 0x06, 0x0c, 0x96, 0x65, 0x18, 0xe4, 0x7f,
+    0x26, 0x44, 0xf3, 0x14, 0xC1, 0xe7, 0xd9, 0x82, 0xf7, 0x64, 0xe8, 0x68,
+    0xf9, 0x6c, 0xa9, 0xe7, 0xd1, 0x9b, 0xac, 0xe1, 0xFd, 0xd8, 0x59, 0xb7,
+    0x8e, 0xdc, 0x24, 0xb8, 0xa7, 0xaf, 0x20, 0xee, 0x6c, 0x61, 0x48, 0x41,
+    0xB4, 0x62, 0x3c, 0xcb, 0x2c, 0xbb, 0xe4, 0x44, 0x97, 0x8a, 0x5e, 0x2f,
+    0x7f, 0x2b, 0x10, 0xcc, 0x7d, 0x89, 0x32, 0xfd, 0xfd, 0x58, 0x7f, 0xd8,
+    0xc7, 0x33, 0xd1, 0x6a, 0xc7, 0xba, 0x78, 0x69,
+};
+
+// Set our expectation & event indications of the interrupts we intend to
+// exercise in this test. These are declared volatile since they are used by the
+// interrupt handler.
+static volatile bool exp_spi_device_irq_rx_full;
+static volatile bool spi_device_irq_rx_full_fired;
+static volatile bool exp_spi_device_irq_rx_above_level;
+static volatile bool spi_device_irq_rx_above_level_fired;
+static volatile bool exp_spi_device_irq_tx_below_level;
+static volatile bool spi_device_irq_tx_below_level_fired;
+static volatile bool exp_spi_device_irq_rx_overflow;
+static volatile bool spi_device_irq_rx_overflow_fired;
+static volatile bool exp_spi_device_irq_tx_underflow;
+static volatile bool spi_device_irq_tx_underflow_fired;
```

Comment:
This big pile of identical-seeming variables seems like a typo hazard.

If you're only expecting one IRQ at a time, you could have two volatile dif_spi_device_irq_t variables; one holding the expected IRQ, and one holding the most recent IRQ the handler saw; this will let you simplify the switch below.

If you're expecting multiple IRQs, you could have two arrays of `bool`s wide enough so that you could do e.g.
```c
CHECK(expected_irqs[my_irq_type]);
fired_irqs[my_irq_type] = true;
```